tests/regression/test_netgen.py

```diff
@@ -0,0 +1,220 @@
+    diff = np.array([poisson(h)[0] for h in [1/2, 1/4, 1/8]])
+    print("l2 error norms:", diff)
+    conv = np.log2(diff[:-1] / diff[1:])
+    print("convergence order:", conv)
+    assert (np.array(conv) > 2.8).all()
+
+
+def test_firedrake_Poisson3D_netgen():
+    diff = np.array([poisson3D(h) for h in [2,1,1/2]])
+    print("l2 error norms:", diff)
+    conv = np.log2(diff[:-1] / diff[1:])
+    print("convergence order:", conv)
+    assert (np.array(conv) > 2.8).all()
+
+
+def test_firedrake_Adaptivity_netgen():
+    gc.collect()
+    comm = MPI.COMM_WORLD
+    from petsc4py import PETSc
+    from slepc4py import SLEPc
+
+    def Solve(msh,labels):
+        V = FunctionSpace(msh, "CG", 2)
+        u = TrialFunction(V)
+        v = TestFunction(V)
+        a = inner(grad(u), grad(v))*dx
+        m = (u*v)*dx
+        uh = Function(V)
+        bc = DirichletBC(V, 0, labels)
+        A = assemble(a, bcs=bc)
+        M = assemble(m)
+        Asc, Msc = A.M.handle, M.M.handle
+        E = SLEPc.EPS().create()
+        E.setType(SLEPc.EPS.Type.ARNOLDI)
+        E.setProblemType(SLEPc.EPS.ProblemType.GHEP)
+        E.setDimensions(1, SLEPc.DECIDE)
+        E.setOperators(Asc, Msc)
+        ST = E.getST()
+        ST.setType(SLEPc.ST.Type.SINVERT)
+        PC = ST.getKSP().getPC()
+        PC.setType("lu")
+        PC.setFactorSolverType("mumps")
+        E.setST(ST)
+        E.solve()
+        vr, vi = Asc.getVecs()
+        with uh.dat.vec_wo as vr:
+            lam = E.getEigenpair(0, vr, vi)
+        return (lam, uh, V)
+
+    def Mark(msh, uh, lam):
+        W = FunctionSpace(msh, "DG", 0)
+        w = TestFunction(W)
+        R_T = lam.real*uh + div(grad(uh))
+        n = FacetNormal(V.mesh())
+        h = CellDiameter(msh)
+        R_dT = dot(grad(uh), n)
+        eta = assemble(h**2*R_T**2*w*dx + (h("+")+h("-"))*(R_dT("+")-R_dT("-"))**2*(w("+")+w("-"))*dS)
+        frac = .95
+        delfrac = .05
+        part = .2
+        with eta.dat.vec as etaVec:
+            markedVec = etaVec.duplicate()
+            markedVec.set(0)
+            sum_eta = etaVec.sum()
+            if sum_eta < tolerance:
+                return markedVec
+            eta_max = etaVec.max()[1]
+            sct, etaVec0 = PETSc.Scatter.toZero(etaVec)
+            markedVec0 = etaVec0.duplicate()
+            sct(etaVec,etaVec0)
+            if etaVec.getComm().getRank() == 0:
+                eta = etaVec0.getArray()
+                marked = np.zeros(eta.size, dtype='bool')
+                sum_marked_eta = 0.
+                while sum_marked_eta < part*sum_eta:
+                    new_marked = (~marked) & (eta > frac*eta_max)
+                    sum_marked_eta += sum(eta[new_marked])
+                    marked += new_marked
+                    frac -= delfrac
+                markedVec0.getArray()[:] = 1.0*marked[:]
+            sct(markedVec0,markedVec,mode=PETSc.Scatter.Mode.REVERSE)
+            return markedVec
+
+    def Refine(msh, marked):
```

Comment:
make this a mesh method. Have `marked` be a DG0 field over the mesh which is 0 or 1.